zio-http/shared/src/main/scala/zio/http/codec/HttpCodec.scala

```diff
@@ -2264,140 +2269,220 @@ object HttpCodec extends ContentCodecs with HeaderCodecs with MethodCodecs with
+          }
+        case s: Schema.Lazy[_]                                                   => StringCodec.fromSchema(s.schema)
+        case s: Schema.Transform[Any, Any, _] @unchecked                         =>
+          val stringCodec = StringCodec.fromSchema(s.schema)
+          new StringCodec[Any] {
+            override def decode(whole: String): Either[DecodeError, Any] =
+              stringCodec.decode(whole).flatMap(s.f(_).left.map(DecodeError.ReadError(Cause.empty, _)))
+
+            override def streamDecoder: ZPipeline[Any, DecodeError, Char, Any] =
+              stringCodec.streamDecoder >>> ZPipeline.map(s.f(_).left.map(DecodeError.ReadError(Cause.empty, _)))
+
+            override def encode(value: Any): String =
+              stringCodec.encode(s.g(value).fold(msg => throw new Exception(msg), identity))
+
+            override def streamEncoder: ZPipeline[Any, Nothing, Any, Char] =
+              ZPipeline.map[Any, Any](
+                s.g(_).fold(msg => throw new Exception(msg), identity),
+              ) >>> stringCodec.streamEncoder
+          }
+        case schema: Schema[_]                                                   => StringCodec.fromSchema(schema)
+      }).asInstanceOf[StringCodec[Any]]
     }
+
+    private def unlazyField(field: Schema.Field[_, _]): Schema.Field[_, _] = field match {
+      case f if f.schema.isInstanceOf[Schema.Lazy[_]] =>
+        Schema.Field(
+          f.name,
+          f.schema.asInstanceOf[Schema.Lazy[_]].schema.asInstanceOf[Schema[Any]],
+          f.annotations,
+          f.validation.asInstanceOf[Validation[Any]],
+          f.get.asInstanceOf[Any => Any],
+          f.set.asInstanceOf[(Any, Any) => Any],
+        )
+      case f                                          => f
+    }
+
+    def validate(value: Any): Chunk[ValidationError] =
+      schema.asInstanceOf[Schema[_]] match {
+        case Schema.Optional(schema: Schema[Any], _) =>
+          schema.validate(value)(schema)
+        case schema: Schema[_]                       =>
+          schema.asInstanceOf[Schema[Any]].validate(value)(schema.asInstanceOf[Schema[Any]])
+      }
+    val defaultValue: A                              =
+      if (schema.isInstanceOf[Schema.Collection[_, _]]) {
+        Try(schema.asInstanceOf[Schema.Collection[A, _]].empty).fold(
+          _ => null.asInstanceOf[A],
+          identity,
+        )
+      } else {
+        schema.defaultValue match {
+          case Right(value) => value
+          case Left(_)      =>
+            schema match {
+              case _: Schema.Optional[_]               => None.asInstanceOf[A]
+              case collection: Schema.Collection[A, _] =>
+                Try(collection.empty).fold(
+                  _ => null.asInstanceOf[A],
+                  identity,
+                )
+              case _                                   => null.asInstanceOf[A]
+            }
+        }
+      }
+
+  }
+
+  object SchemaCodec {
```

Comment:
Ditto about private or public (& locked down) top-level. 

I am not sure `SchemaCodec` is a good name for this so we'll have to do better if you can't make the whole thing private.